File: leetcode/066-plus-one/solution.py
from typing import List


class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        for i in range(len(digits) - 1, -1, -1):
            if digits[i] < 9:
                digits[i] += 1
                return digits
            digits[i] = 0
        return [1] + digits


if __name__ == "__main__":
    print(Solution().plusOne([1, 2, 3]))  # [1, 2, 4]
    print(Solution().plusOne([9, 9]))     # [1, 0, 0]

File: leetcode/066-plus-one/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        for (int i = (int)digits.size() - 1; i >= 0; i--) {
            if (digits[i] < 9) {
                digits[i]++;
                return digits;
            }
            digits[i] = 0;
        }
        digits.insert(digits.begin(), 1);
        return digits;
    }
};

int main() {
    Solution sol;
    vector<int> d1 = {1, 2, 3};
    for (int x : sol.plusOne(d1)) cout << x; // 124
    cout << endl;
    vector<int> d2 = {9, 9};
    for (int x : sol.plusOne(d2)) cout << x; // 100
    cout << endl;
    return 0;
}
